watertap/unit_models/crystallizer.py

```diff
@@ -0,0 +1,696 @@
+            units=pyunits.m**3,
+            doc="Crystallizer minimum active volume, i.e. volume of liquid-solid suspension"
+            )
+
+
+
+        # Add state blocks for inlet, outlet, and waste
+        # These include the state variables and any other properties on demand
+        # Add inlet block
+        tmp_dict = dict(**self.config.property_package_args)
+        tmp_dict["has_phase_equilibrium"] = False
+        tmp_dict["parameters"] = self.config.property_package
+        tmp_dict["defined_state"] = True  # inlet block is an inlet
+        self.properties_in = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,  # time domain for the state block, just 0 in this case
+            doc="Material properties of inlet",
+            default=tmp_dict)
+
+        # Add outlet and waste block
+        tmp_dict["defined_state"] = False  # outlet and waste block is not an inlet
+        self.properties_out = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,
+            doc="Material properties of liquid outlet",
+            default=tmp_dict)
+
+        self.properties_solids = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,
+            doc="Material properties of solid crystals at outlet",
+            default=tmp_dict)
+
+        self.properties_vapor = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,
+            doc="Material properties of water vapour at outlet",
+            default=tmp_dict)
+
+        # Add ports - oftentimes users interact with these rather than the state blocks
+        self.add_port(name='inlet', block=self.properties_in)
+        self.add_port(name='outlet', block=self.properties_out)
+        self.add_port(name='solids', block=self.properties_solids)
+        self.add_port(name='vapor', block=self.properties_vapor)
+
+
+
+        # Add constraints
+        # 1. Material balances
+
+        ## (i) Solute mass balances
+        @self.Constraint(self.config.property_package.component_list,
+                         doc="Mass balance for components")
+        def eq_mass_balance_solutes(b, j):
+            if j in solute_set: 
+                return (b.properties_in[0].flow_mass_phase_comp['Liq', j] + b.properties_in[0].flow_mass_phase_comp['Sol', j] \
+                    == b.properties_out[0].flow_mass_phase_comp['Liq', j] + b.properties_solids[0].flow_mass_phase_comp['Sol', j])
+            else:
+                return Constraint.Skip
+
+        ## (ii) Solvent mass balances
+        @self.Constraint(doc="Mass balance for solvent")
+        def eq_mass_balance_solvent(b):
+                return (b.properties_in[0].flow_mass_phase_comp['Liq', 'H2O'] 
+                    == b.properties_out[0].flow_mass_phase_comp['Liq', 'H2O']
+                    + b.properties_vapor[0].flow_mass_phase_comp['Vap', 'H2O'])               
```

Comment:
I think you can just do the species mass balance and not differentiate between the solute and solvents, and include all of the relevant variables:
A constraint indexed by component could:
```
return sum(b.properties_in[0].flow_mass_phase_comp[p, j] for all p) ==
           sum(b.properties_out[0].flow_mass_phase_comp[p, j] for all p)
           sum(b.properties_vapor[0].flow_mass_phase_comp[p, j] for all p)
           sum(b.properties_solids[0].flow_mass_phase_comp[p, j] for all p)
```
This won't work directly because it doesn't use phase_component_set for the p summation, and I'm not 100% sure how to do that at the moment, but there is probably a way to pull all the relevant phases and sum across them.

This type of formulation would force you to fix the unneeded variables, but I saw that you do that anyway below.

Reply:
@TimBartholomew I have reformulated this as:

```
        @self.Constraint(self.config.property_package.component_list, doc="Mass balance for components")
        def eq_mass_balance_constraints(b, j):
            return (
                sum(b.properties_in[0].flow_mass_phase_comp[p, j] for p in self.config.property_package.phase_list if (p,j) in b.properties_in[0].phase_component_set) == \
                sum(b.properties_out[0].flow_mass_phase_comp[p, j] for p in self.config.property_package.phase_list if (p,j) in b.properties_out[0].phase_component_set)
                + sum(b.properties_vapor[0].flow_mass_phase_comp[p, j] for p in self.config.property_package.phase_list if (p,j) in b.properties_vapor[0].phase_component_set)
                + sum(b.properties_solids[0].flow_mass_phase_comp[p, j] for p in self.config.property_package.phase_list if (p,j) in b.properties_solids[0].phase_component_set) 
           )
```